### ISSUE TITLE
Upgrading datadog-php-tracer causes PHP Fatal Errors until Apache is restarted

### ISSUE THREAD

**Issue by ctrlrsf** [DataDog/dd-trace-php#474]
Hi,

While testing 0.25.0 upgrade to 0.27.1 we saw that as soon as the new datadog-php-tracer_0.27.1_amd64.deb package is installed, Apache / PHP will throw PHP Fatal Errors until Apache is restarted. It makes sense since some of the underlying C library has changed, but not sure this is documented anywhere and if not handled properly will bring down a site since all requests involving PHP return 500 (until restart). For now we're handling this by pinning datadog-php-tracer to a version with config management, and restarting Apache immediately when/if it changes.

Example:
1. Start with datadog-php-tracer_0.25.0_amd64.deb installed
1. Make test requests (200 as expected)
1. Install datadog-php-tracer_0.27.1_amd64.deb
1. Make test requests (will return 500)
1. Start seeing PHP Fatal Errors (errors below)
1. Restart apache2, errors are gone

Fatal error examples:
0.25.0 to 0.27.1:
```
PHP Fatal error:  Uncaught Error: Call to undefined function DDTrace\\dd_trace_generate_id() in /opt/datadog-php/dd-trace-sources/src/DDTrace/SpanContext.php:41\nStack trace:\n#0 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(184): DDTrace\\SpanContext::createAsRoot()\n#1 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(250): DDTrace\\Tracer->startSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#2 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(220): DDTrace\\Tracer->startActiveSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#3 /opt/datadog-php/dd-trace-sources/src/DDTrace/Bootstrap.php(96): DDTrace\\Tracer->startRootSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#4 /opt/datadog-php/dd-trace-sources/src/DDTrace/Bootstrap.php(30): DDTrace\\Bootstrap::initRootSpan()\n#5 /opt/datadog-php/dd-trace-sources/bridge/dd_init.php(14): DDTrace\\Bootstrap::tracerOnce()\n#6 /opt/datadog-php/dd-trace-sources/bridge/functions.php(139): require('/opt/datadog-ph...')\n#7 /var/www/xyz/releases/xyz in /opt/datadog-php/dd-trace-sources/src/DDTrace/SpanContext.php on line 41
```

0.18.0 to 0.27.2:
```
PHP Fatal error:  Uncaught Error: Class 'DDTrace\\ID' not found in /opt/datadog-php/dd-trace-sources/src/DDTrace/SpanContext.php:85\nStack trace:\n#0 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(156): DDTrace\\SpanContext::createAsRoot()\n#1 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(220): DDTrace\\Tracer->startSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#2 /opt/datadog-php/dd-trace-sources/src/DDTrace/Tracer.php(190): DDTrace\\Tracer->startActiveSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#3 /opt/datadog-php/dd-trace-sources/src/DDTrace/Bootstrap.php(99): DDTrace\\Tracer->startRootSpan('web.request', Object(DDTrace\\StartSpanOptions))\n#4 /opt/datadog-php/dd-trace-sources/src/DDTrace/Bootstrap.php(33): DDTrace\\Bootstrap::initRootSpan()\n#5 /opt/datadog-php/dd-trace-sources/bridge/dd_init.php(14): DDTrace\\Bootstrap::tracerOnce()\n#6 /opt/datadog-php/dd-trace-sources/bridge/functions.php(127): require('/opt/datadog-ph...')\n#7 /var/www/xyz/releases/xyz/app/Ve in /opt/datadog-php/dd-trace-sources/src/DDTrace/SpanContext.php on line 85

PHP Fatal error:  Uncaught Error: Call to undefined function DDTrace\\Bridge\\dd_trace_forward_call() in /opt/datadog-php/dd-trace-sources/bridge/functions.php:116\nStack trace:\n#0 /var/www/xyz/releases/xyz/app/Vendor/cakephp/cakephp/lib/Cake/bootstrap.php(142): DDTrace\\Bridge\\{closure}(Array)\n#1 /var/www/xyz/releases/xyz/app/webroot/index.php(105): include('/var/www/xyz/re...')\n#2 {main}\n  thrown in /opt/datadog-php/dd-trace-sources/bridge/functions.php on line 116

PHP Fatal error:  Uncaught Error: Call to undefined function DDTrace\\Bridge\\dd_trace_forward_call() in /opt/datadog-php/dd-trace-sources/bridge/functions.php:116\nStack trace:\n#0 /var/www/xyz/releases/xyz/app/Vendor/cakephp/cakephp/lib/Cake/bootstrap.php(142): DDTrace\\Bridge\\{closure}(Array)\n#1 /var/www/xyz/releases/xyz/app/webroot/index.php(105): include('/var/www/xyz/re...')\n#2 {main}\n  thrown in /opt/datadog-php/dd-trace-sources/bridge/functions.php on line 116
```

**Comment by ctrlrsf** [comment]
Seems like opcache was caching scripts in /opt/datadog-php/ so Apache/PHP would keep running older PHP code against the new C library.

Another workaround is to add /opt/datadog-php to opcache blacklist file.

1. Add `/opt/datadog-php/` to `/etc/php/opcache_blacklist`
1. Set `opcache.blacklist_filename=/etc/php/opcache_blacklist` in `/etc/php/7.2/apache2/php.ini`

Snip of opcache stats showing /opt/datadog-php/ scripts in cache:
```
  "scripts": {
    "/opt/datadog-php/dd-trace-sources/src/DDTrace/Propagators/TextMap.php": {
      "full_path": "/opt/datadog-php/dd-trace-sources/src/DDTrace/Propagators/TextMap.php",
      "hits": 55,
      "memory_consumption": 9432,
      "last_used": "Thu Jun  6 18:18:16 2019",
      "last_used_timestamp": 1559845096,
      "timestamp": 1559834211
    },
    "/opt/datadog-php/dd-trace-sources/src/DDTrace/ScopeManager.php": {
      "full_path": "/opt/datadog-php/dd-trace-sources/src/DDTrace/ScopeManager.php",
      "hits": 55,
      "memory_consumption": 5816,
      "last_used": "Thu Jun  6 18:18:16 2019",
      "last_used_timestamp": 1559845096,
      "timestamp": 1559834211
    },
```

**Comment by pawelchcki** [comment]
Thanks fro reporting @ctrlrsf! Its possible that the extension actually wasn't properly updated in the second case you've mentioned. Checking PHP -i e.g. by running `php -i | grep "Datadog PHP tracer" -A5` could confirm that.
We're internally thinking about adding a version check to ensure that extension version matches

**Comment by ctrlrsf** [comment]
Hi @pawelchcki,

I confirmed the extension is installed for second case. In my test env, I can reproduce it 100% by downgrading and upgrading packages while making test requests.

**Comment by ctrlrsf** [comment]
Hi @SammyK, do you think https://github.com/DataDog/dd-trace-php/pull/488 fixes this?

**Comment by SammyK** [comment]
👋 @ctrlrsf! The changes in #488 are targeted at new installs of the tracer. After upgrading to the latest version of the tracer from an old tracer version before `0.29.0`, the PHP and C code could still get out of sync after upgrading and before restarting. But after upgrading to the latest tracer version (starting with `0.29.0`) and restarting the PHP SAPI, the tracer will temporality disable itself every time the tracer is upgraded and will re-enable itself after the PHP SAPI is restarted.

**TL;DR:** Yes for upgrades that occur after the tracer is upgraded to at least version `0.29.0`. :)

**Comment by brandonburkett** [comment]
Is there documentation on how to update tracer?  (EX: install RPM, restart service).  I saw install documentation, but it wasn't clear on the update process.

**Comment by SammyK** [comment]
Hey @ctrlrsf! I'm closing this issue since this issue should be fixed. Please feel free to reopen if you're still having trouble.

Hey @brandonburkett. I'm sorry we missed your reply. The trace documentation has an [upgrading section](https://docs.datadoghq.com/tracing/setup/php/#upgrading).
